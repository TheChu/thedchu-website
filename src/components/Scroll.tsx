import React, {
  cloneElement,
  Component,
  KeyboardEvent,
  MouseEvent,
  ReactElement,
} from 'react';
import smoothscroll from 'smoothscroll-polyfill';

const Element = (props) => props.children;

interface ScrollProps {
  children: ReactElement;
  element?: string;
  offset?: number;
  onClick?: (event: KeyboardEvent | MouseEvent) => void;
  timeout?: number;
  type?: string;
}

class Scroll extends Component<ScrollProps> {
  constructor(props: ScrollProps) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  componentDidMount(): void {
    smoothscroll.polyfill();
  }

  handleClick = (event: KeyboardEvent | MouseEvent): void => {
    event.preventDefault();
    const { onClick = () => {} } = this.props; // eslint-disable-line
    let elem: Element;
    let scroll = true;
    const { type, element, offset, timeout } = this.props;
    if (type && element) {
      switch (type) {
        case 'class':
          [elem] = document.getElementsByClassName(element);
          scroll = !!elem;
          break;
        case 'id':
          elem = document.getElementById(element);
          scroll = !!elem;
          break;
        default:
      }
    }

    if (scroll) {
      this.scrollTo(elem, offset, timeout);
    } else {
      console.log(`Element not found: ${element}`); // eslint-disable-line
    }

    onClick(event);
  };

  scrollTo = (element: Element, offSet = 0, timeout = null): void => {
    const elemPos = element
      ? element.getBoundingClientRect().top + window.pageYOffset
      : 0;
    if (timeout) {
      setTimeout(() => {
        window.scroll({ top: elemPos + offSet, left: 0, behavior: 'smooth' });
      }, timeout);
    } else {
      window.scroll({ top: elemPos + offSet, left: 0, behavior: 'smooth' });
    }
  };

  render(): ReactElement {
    const { children } = this.props;

    return (
      <Element>
        {cloneElement(children, {
          onClick: this.handleClick,
        })}
      </Element>
    );
  }
}

export default Scroll;
