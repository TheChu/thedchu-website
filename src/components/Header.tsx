import { Link } from 'gatsby';
import React, { Component, ReactElement } from 'react';
import config from '../../config';
import Scroll from './Scroll';
import logo from '../assets/img/thedchu-logo.svg';

interface HeaderState {
  openMenu: boolean;
  visibilityClass: string;
}

export default class Header extends Component<{}, HeaderState> {
  constructor(props: {}) {
    super(props);
    this.state = {
      openMenu: false,
      visibilityClass: '',
    };
  }

  componentDidMount(): void {
    window.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount(): void {
    window.removeEventListener('scroll', this.handleScroll);
  }

  toggleMenu = (value: boolean): void => {
    this.setState({ openMenu: value });
  };

  handleScroll = (): void => {
    const { visibilityClass } = this.state;
    if (window.pageYOffset > 300) {
      if (visibilityClass !== 'navbar-scrolled') {
        this.setState({ visibilityClass: 'navbar-scrolled' });
      }
    } else if (visibilityClass === 'navbar-scrolled') {
      this.setState({ visibilityClass: '' });
    }
  };

  render(): ReactElement {
    const { openMenu, visibilityClass } = this.state;
    return (
      <nav
        className={`navbar navbar-expand-lg navbar-light fixed-top py-3 ${visibilityClass}`}
        id="mainNav"
      >
        <div className="container">
          <div className="navbar-item">
            <ul className="navbar-nav mr-auto my-2 my-lg-0">
              {config.socialLinks.map((social) => {
                const { icon, name, style, url } = social;
                return (
                  <li key={url} className="nav-item">
                    <a href={url} className="nav-link nav-link-icon mx-2">
                      <i className={`fa${style} ${icon}`} aria-label={name} />
                    </a>
                  </li>
                );
              })}
            </ul>
          </div>
          <Link to="/" className="navbar-brand navbar-item">
            <svg
              className="navbar-brand-logo"
              viewBox="0 0 1000 512"
              aria-label="DChu"
            >
              <use href={`${logo}#thedchu-logo`} />
            </svg>
          </Link>

          <div className="navbar-toggler-container">
            <button
              onClick={() => this.toggleMenu(!openMenu)}
              className={`navbar-toggler navbar-toggler-right ${
                openMenu ? '' : 'collapsed'
              }`}
              type="button"
              aria-controls="navbarResponsive"
              aria-expanded={openMenu}
              aria-label="Toggle navigation"
            >
              <span className="navbar-toggler-icon" />
            </button>
          </div>

          <div
            className={`collapse navbar-collapse ${openMenu ? 'show' : ''}`}
            id="navbarResponsive"
          >
            <ul className="navbar-nav ml-auto my-2 my-lg-0">
              <li className="nav-item">
                <Scroll
                  onClick={() => this.toggleMenu(!openMenu)}
                  type="id"
                  element="about"
                >
                  <a className="nav-link" href="#about">
                    About
                  </a>
                </Scroll>
              </li>
              <li className="nav-item">
                <Scroll
                  onClick={() => this.toggleMenu(!openMenu)}
                  type="id"
                  element="photos"
                >
                  <a className="nav-link" href="#photos">
                    Photos
                  </a>
                </Scroll>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    );
  }
}
