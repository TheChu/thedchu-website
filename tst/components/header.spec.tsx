import { fireEvent, render } from '@testing-library/react';
import React, { cloneElement } from 'react';
import Header from '../../src/components/Header';
import Scroll from '../../src/components/Scroll';

jest.mock('../../src/components/Scroll', () => jest.fn());
(Scroll as jest.Mock).mockImplementation(({ children, onClick }) => (
  <>{cloneElement(children, { onClick })}</>
));

describe('Header', () => {
  it('should match Header snapshot', () => {
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the correct number of links', () => {
    const { getAllByRole } = render(<Header />);
    expect(getAllByRole('link').length).toBe(5);
  });

  const navigationLinks = [
    { text: 'About', location: '#about' },
    { text: 'Services', location: '#services' },
    { text: 'Portfolio', location: '#portfolio' },
    { text: 'Contact', location: '#contact' },
  ];

  it.each(navigationLinks)('should render %s link', (link) => {
    const { getByRole } = render(<Header />);
    expect(getByRole('link', { name: link.text })).toHaveAttribute(
      'href',
      link.location
    );
  });

  it('should render site title', () => {
    const { getByRole } = render(<Header />);
    expect(getByRole('link', { name: 'thedchu' })).toHaveAttribute(
      'href',
      '#page-top'
    );
  });

  it('should change navigation bar when scrolling past threshold', () => {
    const { getByRole } = render(<Header />);
    expect(getByRole('navigation')).not.toHaveClass('navbar-scrolled');

    // Past threshold of 300
    fireEvent.scroll(window, { target: { pageYOffset: 500 } });
    expect(getByRole('navigation')).toHaveClass('navbar-scrolled');

    fireEvent.scroll(window, { target: { pageYOffset: 600 } });
    expect(getByRole('navigation')).toHaveClass('navbar-scrolled');

    // Within threshold of 300
    fireEvent.scroll(window, { target: { pageYOffset: 0 } });
    expect(getByRole('navigation')).not.toHaveClass('navbar-scrolled');

    fireEvent.scroll(window, { target: { pageYOffset: 100 } });
    expect(getByRole('navigation')).not.toHaveClass('navbar-scrolled');
  });

  // The navigation bar expands and closes only on mobile (or a small desktop
  // window) while on desktop it is always shown and cannot be closed. Although
  // the following tests technically only apply for mobile, the button does
  // exist in the test environment DOM and has a `aria-expanded=false` at
  // startup. So, we can test funcionality without needing to inject styles.
  // This is now tested more "correctly" with Cypress.
  describe('mobile navigation bar menu', () => {
    const assertNavMenuClosed = (getByRole, queryByRole) => {
      expect(
        getByRole('button', { name: 'Toggle navigation', expanded: false })
      ).toBeInTheDocument();
      expect(
        queryByRole('button', { name: 'Toggle navigation', expanded: true })
      ).not.toBeInTheDocument();
    };

    const assertNavMenuOpen = (getByRole, queryByRole) => {
      expect(
        queryByRole('button', { name: 'Toggle navigation', expanded: false })
      ).not.toBeInTheDocument();
      expect(
        getByRole('button', { name: 'Toggle navigation', expanded: true })
      ).toBeInTheDocument();
    };

    it('should display and hide navigation menu when clicking `Toggle navigation` button', () => {
      const { getByRole, queryByRole } = render(<Header />);
      assertNavMenuClosed(getByRole, queryByRole);

      fireEvent.click(getByRole('button', { name: 'Toggle navigation' }));
      assertNavMenuOpen(getByRole, queryByRole);

      fireEvent.click(getByRole('button', { name: 'Toggle navigation' }));
      assertNavMenuClosed(getByRole, queryByRole);
    });

    it.each(navigationLinks)(
      'should hide navigation menu when clicking %s link',
      (link) => {
        const { getByRole, queryByRole } = render(<Header />);
        assertNavMenuClosed(getByRole, queryByRole);

        fireEvent.click(getByRole('button', { name: 'Toggle navigation' }));
        assertNavMenuOpen(getByRole, queryByRole);

        fireEvent.click(getByRole('link', { name: link.text }));
        assertNavMenuClosed(getByRole, queryByRole);
      }
    );
  });
});
