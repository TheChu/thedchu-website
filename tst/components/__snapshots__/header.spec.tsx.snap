// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match Header snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-light fixed-top py-3 "
    id="mainNav"
  >
    <div
      class="container"
    >
      <div
        class="navbar-item"
      >
        <ul
          class="navbar-nav mr-auto my-2 my-lg-0"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link nav-link-icon mx-2"
              href="https://instagram.com/thedchu"
            >
              <i
                aria-label="Instagram"
                class="fab fa-instagram"
              />
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link nav-link-icon mx-2"
              href="https://linkedin.com/in/thedchu"
            >
              <i
                aria-label="LinkedIn"
                class="fab fa-linkedin-in"
              />
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link nav-link-icon mx-2"
              href="https://github.com/thechu"
            >
              <i
                aria-label="Github"
                class="fab fa-github"
              />
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link nav-link-icon mx-2"
              href="mailto:chu.david.a@gmail.com"
            >
              <i
                aria-label="Email"
                class="fas fa-envelope"
              />
            </a>
          </li>
        </ul>
      </div>
      <a
        class="navbar-brand navbar-item"
        href="/"
      >
        <svg
          aria-label="DChu"
          class="navbar-brand-logo"
          viewBox="0 0 1000 512"
        >
          <use
            href="test-file-stub#thedchu-logo"
          />
        </svg>
      </a>
      <div
        class="navbar-toggler-container"
      >
        <button
          aria-controls="navbarResponsive"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler navbar-toggler-right collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
      </div>
      <div
        class="collapse navbar-collapse "
        id="navbarResponsive"
      >
        <ul
          class="navbar-nav ml-auto my-2 my-lg-0"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#about"
            >
              About
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#photos"
            >
              Photos
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
